File: App/infra/lambda/state-handler.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand, PutCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyEventV2WithJWTAuthorizer, APIGatewayProxyResultV2 } from 'aws-lambda';

const ddb = DynamoDBDocumentClient.from(new DynamoDBClient({}));
const TABLE_NAME = process.env.TABLE_NAME!;

// LearnerState blobs are a few hundred KB at absolute most; cap well below
// the DynamoDB 400 KB item limit to fail fast on garbage payloads.
const MAX_STATE_BYTES = 380_000;

function json(statusCode: number, body: unknown): APIGatewayProxyResultV2 {
  return {
    statusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

export async function handler(
  event: APIGatewayProxyEventV2WithJWTAuthorizer,
): Promise<APIGatewayProxyResultV2> {
  // JWT authorizer guarantees a validated token; `sub` is the stable user id.
  const userId = event.requestContext.authorizer?.jwt?.claims?.sub;
  if (typeof userId !== 'string' || !userId) {
    return json(401, { error: 'unauthorized' });
  }

  const method = event.requestContext.http.method;

  if (method === 'GET') {
    const res = await ddb.send(
      new GetCommand({ TableName: TABLE_NAME, Key: { userId } }),
    );
    if (!res.Item) return json(404, { error: 'no state saved yet' });
    return json(200, { state: res.Item.state, updatedAt: res.Item.updatedAt });
  }

  if (method === 'PUT') {
    if (!event.body) return json(400, { error: 'missing body' });
    if (Buffer.byteLength(event.body, 'utf8') > MAX_STATE_BYTES) {
      return json(413, { error: 'state too large' });
    }
    let parsed: { state?: unknown };
    try {
      parsed = JSON.parse(event.body);
    } catch {
      return json(400, { error: 'invalid JSON' });
    }
    const state = parsed.state;
    if (
      typeof state !== 'object' ||
      state === null ||
      typeof (state as { version?: unknown }).version !== 'number'
    ) {
      return json(400, { error: 'body must be { state: LearnerState }' });
    }
    const updatedAt = new Date().toISOString();
    await ddb.send(
      new PutCommand({
        TableName: TABLE_NAME,
        Item: { userId, state, updatedAt },
      }),
    );
    return json(200, { updatedAt });
  }

  return json(405, { error: 'method not allowed' });
}

File: App/infra/lambda/state_handler.py
"""State-sync Lambda handler (Python port of state-handler.ts).

Serves `GET /state` and `PUT /state` behind an API Gateway HTTP API with a
Cognito JWT authorizer. One DynamoDB item per user holds the whole
LearnerState blob, keyed by the token's `sub` claim.

Behaviour is a faithful port of the original TypeScript handler:
  GET  -> 200 { state, updatedAt } | 404 { error }
  PUT  -> 200 { updatedAt } | 400 | 413
  missing/invalid token -> 401
  any other method       -> 405
"""

import json
import os
from datetime import datetime, timezone

import boto3

# Reuse the client across warm invocations.
_TABLE_NAME = os.environ["TABLE_NAME"]
_table = boto3.resource("dynamodb").Table(_TABLE_NAME)

# LearnerState blobs are a few hundred KB at absolute most; cap well below
# the DynamoDB 400 KB item limit to fail fast on garbage payloads.
MAX_STATE_BYTES = 380_000


def _json(status_code, body):
    return {
        "statusCode": status_code,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps(body),
    }


def _user_id(event):
    """Pull the validated `sub` claim the JWT authorizer attached to the event."""
    try:
        claims = event["requestContext"]["authorizer"]["jwt"]["claims"]
    except (KeyError, TypeError):
        return None
    sub = claims.get("sub") if isinstance(claims, dict) else None
    return sub if isinstance(sub, str) and sub else None


def handler(event, _context=None):
    user_id = _user_id(event)
    if not user_id:
        return _json(401, {"error": "unauthorized"})

    method = event.get("requestContext", {}).get("http", {}).get("method")

    if method == "GET":
        res = _table.get_item(Key={"userId": user_id})
        item = res.get("Item")
        if not item:
            return _json(404, {"error": "no state saved yet"})
        return _json(200, {"state": item.get("state"), "updatedAt": item.get("updatedAt")})

    if method == "PUT":
        body = event.get("body")
        if not body:
            return _json(400, {"error": "missing body"})
        # API Gateway may base64-encode the body; the SPA sends plain JSON, but
        # measure the raw bytes the client sent to match the original cap.
        if len(body.encode("utf-8")) > MAX_STATE_BYTES:
            return _json(413, {"error": "state too large"})
        try:
            parsed = json.loads(body)
        except (ValueError, TypeError):
            return _json(400, {"error": "invalid JSON"})

        state = parsed.get("state") if isinstance(parsed, dict) else None
        if not isinstance(state, dict) or not isinstance(state.get("version"), (int, float)) \
                or isinstance(state.get("version"), bool):
            return _json(400, {"error": "body must be { state: LearnerState }"})

        updated_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
        _table.put_item(Item={"userId": user_id, "state": state, "updatedAt": updated_at})
        return _json(200, {"updatedAt": updated_at})

    return _json(405, {"error": "method not allowed"})
